polaris-core/src/main/java/org/apache/polaris/core/persistence/IntegrationPersistence.java

```diff
@@ -0,0 +1,129 @@
+import org.apache.polaris.core.entity.PolarisPrincipalSecrets;
+import org.apache.polaris.core.storage.PolarisStorageConfigurationInfo;
+import org.apache.polaris.core.storage.PolarisStorageIntegration;
+
+/**
+ * Interface for the necessary "peripheral integration" objects that are logically attached to core
+ * persistence entities but which typically involve additional separate external integrations
+ * related to identity/auth, kms/secrets storage, etc.
+ *
+ * <p>Implementations should orchestrate any necessary multi-phase protocols such as leasing an
+ * external resource before committing a reference to the external resource in the Polaris
+ * persistence layer, etc.
+ */
+public interface IntegrationPersistence {
+  /**
+   * Allows to retrieve to the secrets of a principal given its unique client id
+   *
+   * @param callCtx call context
+   * @param clientId principal client id
+   * @return the secrets
+   */
+  @Nullable
+  PolarisPrincipalSecrets loadPrincipalSecrets(
+      @Nonnull PolarisCallContext callCtx, @Nonnull String clientId);
+
+  /**
+   * generate and store a client id and associated secrets for a newly created principal entity
+   *
+   * @param callCtx call context
+   * @param principalName name of the principal
+   * @param principalId principal id
+   */
+  @Nonnull
+  PolarisPrincipalSecrets generateNewPrincipalSecrets(
+      @Nonnull PolarisCallContext callCtx, @Nonnull String principalName, long principalId);
+
+  /**
+   * Rotate the secrets of a principal entity, i.e. make the specified main secrets the secondary
+   * and generate a new main secret
+   *
+   * @param callCtx call context
+   * @param clientId principal client id
+   * @param principalId principal id
+   * @param reset true if the principal secrets should be disabled and replaced with a one-time
+   *     password
+   * @param oldSecretHash the principal secret's old main secret hash
+   */
+  @Nullable
+  PolarisPrincipalSecrets rotatePrincipalSecrets(
+      @Nonnull PolarisCallContext callCtx,
+      @Nonnull String clientId,
+      long principalId,
+      boolean reset,
+      @Nonnull String oldSecretHash);
+
+  /**
+   * When dropping a principal, we also need to drop the secrets of that principal
+   *
+   * @param callCtx the call context
+   * @param clientId principal client id
+   * @param principalId the id of the principal whose secrets are dropped
+   */
+  void deletePrincipalSecrets(
+      @Nonnull PolarisCallContext callCtx, @Nonnull String clientId, long principalId);
+
+  /**
+   * Create an in-memory storage integration
+   *
+   * @param callCtx the polaris calllctx
+   * @param catalogId the catalog id
+   * @param entityId the entity id
+   * @param polarisStorageConfigurationInfo the storage configuration information
+   * @return a storage integration object
+   */
+  @Nullable
+  <T extends PolarisStorageConfigurationInfo> PolarisStorageIntegration<T> createStorageIntegration(
```

Comment:
Can `PolarisStorageConfigurationInfo` be an simple entity in `BasePersistence`?

Reply:
Yeah, that was actually one of the original designs, and the current Catalog nesting was thought of as a "simplification" to define "inline StorageConfigurationInfo".

In theory they can be first-class entities and we can have catalogs take referential ids/uris to the configuration info.

Reply:
Why does persistence need to know about object-storage at all? I think persistence and object-storage are very different things. This function and `loadPolarisStorageIntegration` are factories for object-storage specific functionality.

Reply:
Yeah, we might be able to lift these one or two layers higher up in followups.